### PR TITLE
added serializer

### PR DESCRIPTION
**Pull request by nikisalli**

### CODE REVIEW

File: nightmare_firmware/nightmare_firmware.ino [hunk 1]
```diff
@@ -22,6 +24,7 @@ void setup(){
   	digitalWrite(FAN2_PIN, LOW);
 
 	servo::init(SERVO_PIN_TX_ENB, SERVO_PIN_RX_ENB); //init servo circuitry
+	hservo.setPeriodHertz(50);
```
Comment (Cogno-Marco) [inline nightmare_firmware/nightmare_firmware.ino]: maybe would be better to have a const somewhere

File: nightmare_firmware/nightmare_firmware.ino [hunk 2]
```diff
@@ -87,8 +182,17 @@ void loop(){
...
 	}
 
+	if(buf[53] == 1){
+		hservo.attach(SERVO_PIN_HEAD_TILT, SERVO_MIN_WIDTH_HEAD_TILT, SERVO_MAX_WIDTH_HEAD_TILT);
+	} else {
+		hservo.detach();
+	}
+	if(hservo.attached()){
+		hservo.write(buf[26]);
+	}
```
Comment (Cogno-Marco) [inline nightmare_firmware/nightmare_firmware.ino]: a comment on what this does?

File: nightmare_firmware/nightmare_firmware.ino [hunk 3]
```diff
@@ -39,17 +42,109 @@ void setup(){
...
+		buf[13] = 0; //TODO joystick LY
+		buf[14] = 0; //TODO joystick RX
+		buf[15] = 0; //TODO joystick RY
+		buf[16] = 0; //TODO joystick state
+		
+		// calculate checksum
+		int checksum = 0;
+		for(int i=0; i<17; i++){
+			checksum += buf[i];
+		}
```
Comment (Cogno-Marco) [inline nightmare_firmware/nightmare_firmware.ino]:
```suggestion
		for(auto& buf_byte : buf){
			checksum += buf_byte;
		}
```
right? buf is declared some rows above with 18 cells, for goes from 0 to 17, 18 total.

**Review by Cogno-Marco — CHANGES_REQUESTED** [review]
Only minor stuff, pretty good  👍

File: nightmare_firmware/nightmare_firmware.ino (hunk 1 above)
Comment (nikisalli) [reply nightmare_firmware/nightmare_firmware.ino]: fixed in upcoming push

File: nightmare_firmware/nightmare_firmware.ino (hunk 2 above)
Comment (nikisalli) [reply nightmare_firmware/nightmare_firmware.ino]: added comments in upcoming push

File: nightmare_firmware/nightmare_firmware.ino (hunk 3 above)
Comment (nikisalli) [reply nightmare_firmware/nightmare_firmware.ino]: fixed in upcoming push

**Review by Cogno-Marco — APPROVED** [review]
Nice work 👍

File: nightmare_firmware/nightmare_firmware.ino [hunk 4]
```diff
@@ -39,17 +42,109 @@ void setup(){
...
+		buf[15] = 0; //TODO joystick RY
+		buf[16] = 0; //TODO joystick state
+		
+		// calculate checksum
+		int checksum = 0;
+		for(auto& buf_byte : buf){
+			checksum += buf_byte;
+		}
+		
+		buf[17] = (uint8_t)checksum;
```
Comment (chqoot-vgxu) [inline nightmare_firmware/nightmare_firmware.ino:155]: Shouldn't this be `buf[17] = ~(uint8_t)checksum;` ?

File: nightmare_firmware/nightmare_firmware.ino [hunk 5]
```diff
@@ -39,17 +42,109 @@ void setup(){
...
+		>JOYSTICK POS -> 4 bytes describing the two joystick positions in the following order: LX LY RX RY
+		>JOYSTICK STATE -> 1 byte sent by the nightcontrol app through 433Mhz
+		>CHECKSUM -> the first 8 bits of the sum of all the packet's bytes
+ */
+void write_packet(){
+	/* check if all servos are attached */
+	bool all_active = true;
+	for(auto& servo : servos){
+		if(servo.active == false){
+			all_active = false;
```
Comment (chqoot-vgxu) [inline nightmare_firmware/nightmare_firmware.ino:117]: You can break out of this loop, when a non active servo is found

File: nightmare_firmware/nightmare_firmware.ino (hunk 4 above)
Comment (nikisalli) [reply nightmare_firmware/nightmare_firmware.ino:155]: no the ~ comes from the servo protocol that uses the inverted checksum here I'm simply using the sum for simplicity

File: nightmare_firmware/nightmare_firmware.ino (hunk 5 above)
Comment (nikisalli) [reply nightmare_firmware/nightmare_firmware.ino:117]: sry didn't notice, fixed in upcoming push

**Pull request closed without merging**